### PR TITLE
Add gradient checkpointing to Whisper Flax

### PR DESCRIPTION
**Pull request by versae**
It uses `flax.linen.remat` and follows on PRs #13657 and #17994.

# What does this PR do?

Adds gradient_checkpointing to Flax Whisper models.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@sanchit-gandhi @peregilk

### CODE REVIEW

File: src/transformers/models/whisper/modeling_flax_whisper.py [hunk 1]
```diff
@@ -387,16 +390,24 @@ def __call__(
         return outputs
 
 
-# Copied from transformers.models.mbart.modeling_flax_mbart.FlaxMBartEncoderLayerCollection with MBart->Whisper
+# Adapted from transformers.models.mbart.modeling_flax_mbart.FlaxMBartEncoderLayerCollection with MBart->Whisper
```
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/modeling_flax_whisper.py:393]: Fine to remove this comment since we're now not copying directly from MBart!

File: src/transformers/models/whisper/modeling_flax_whisper.py [hunk 2]
```diff
@@ -531,16 +542,24 @@ def __call__(
         return outputs
 
 
-# Copied from transformers.models.mbart.modeling_flax_mbart.FlaxMBartDecoderLayerCollection with MBart->Whisper
+# Adapted from transformers.models.mbart.modeling_flax_mbart.FlaxMBartDecoderLayerCollection with MBart->Whisper
```
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/modeling_flax_whisper.py:545]: Here too!

File: src/transformers/models/whisper/modeling_flax_whisper.py [hunk 3]
```diff
@@ -570,12 +589,12 @@ def __call__(
             else:
                 layer_outputs = decoder_layer(
                     hidden_states,
-                    attention_mask=attention_mask,
-                    encoder_hidden_states=encoder_hidden_states,
-                    encoder_attention_mask=encoder_attention_mask,
-                    init_cache=init_cache,
-                    output_attentions=output_attentions,
-                    deterministic=deterministic,
+                    attention_mask,
```
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/modeling_flax_whisper.py:592]: Note to reviewer: `remat` does not support key-word arguments, hence the need to change to pure arguments

File: src/transformers/models/whisper/modeling_flax_whisper.py [hunk 4]
```diff
@@ -821,11 +850,21 @@ def __init__(
         seed: int = 0,
         dtype: jnp.dtype = jnp.float32,
         _do_init: bool = True,
+        gradient_checkpointing: bool = False,
         **kwargs,
     ):
-        module = self.module_class(config=config, dtype=dtype, **kwargs)
+        self.gradient_checkpointing = gradient_checkpointing
```
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/modeling_flax_whisper.py:856]: No need to set this as an attribute in the Flax Pre-Trained model class - it only needs to be set as an attribute in the nn.Module 
```suggestion

```

File: src/transformers/models/whisper/modeling_flax_whisper.py [hunk 5]
```diff
@@ -821,11 +850,21 @@ def __init__(
...
+        module = self.module_class(config=config, dtype=dtype, gradient_checkpointing=gradient_checkpointing, **kwargs)
         super().__init__(config, module, input_shape=input_shape, seed=seed, dtype=dtype, _do_init=_do_init)
 
+    def enable_gradient_checkpointing(self):
+        self.gradient_checkpointing = True
```
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/modeling_flax_whisper.py:861]: Also here, fine to just set this attribute directly for the nn.Module!
```suggestion

```

File: src/transformers/models/whisper/modeling_flax_whisper.py [hunk 6]
```diff
@@ -821,11 +850,21 @@ def __init__(
...
+        self._module = self.module_class(
+            config=self.config,
+            dtype=self.dtype,
+            gradient_checkpointing=self.gradient_checkpointing,
```
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/modeling_flax_whisper.py:865]: Then this needs to be updated:
```suggestion
            gradient_checkpointing=True,
```

**Review by sanchit-gandhi — APPROVED** [review]
Very nice @versae! Just a few minor nits, but otherwise this PR is looking good!

**Comment by versae** [comment]
Thanks for the review, @sanchit-gandhi! Should be all good now 😃.

**Comment by sanchit-gandhi** [comment]
Amazing @versae! Requesting final review before we can get this merged 🤗

**Review by sgugger — APPROVED** [review]
Thanks for your contribution!

**Comment by versae** [comment]
Thank you! I learnt a lot 🤓
